### ISSUE TITLE
"raft closed timestamp regression" when bringing up force-killed nodes in a region

### ISSUE THREAD

**Issue by otan** [cockroachdb/cockroach#66512]
In a setup I am doing to perform a DRT, I am running TPC-C on a multi-region cluster with multi-region'd tables on `SURVIVE REGION`.

Every 3 minutes, I bring down an entire region of nodes by force killing them, then bringing them back up after 3 minutes.

Sometimes, when bringing up the nodes, we hit this on another node. This case was nodes 1-3 being brought back up, but node 7 running into this assert.
```
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013  ???: raft closed timestamp regression in cmd: "\\\xe5\xf8\x16\b\x8e\xfc\xd7" (term: 21, index: 243); batch state: 1623820438.230255107,0?, command: 1623820437.692808095,0, lease: repl=(n1,s1):1 seq=11 start=1623820437.56
3056021,0 epo=10 pro=1623820437.562975998,0, req: <unknown; not leaseholder>, applying at LAI: 184.
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +(1) assertion failure
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +Wraps: (2) attached stack trace
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  -- stack trace:
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaAppBatch).assertNoCmdClosedTimestampRegression
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  |     /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_application_state_machine.go:1099
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaAppBatch).Stage
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  |     /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_application_state_machine.go:465
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply.mapCmdIter
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  |     /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply/cmd.go:175
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply.(*Task).applyOneBatch
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  |     /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply/task.go:280
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply.(*Task).ApplyCommittedEntries
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  |     /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/apply/task.go:247
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReadyRaftMuLocked
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  |     /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:839
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).handleRaftReady
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  |     /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/replica_raft.go:487
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).processReady
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  |     /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_raft.go:515
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*raftScheduler).worker
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  |     /go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/scheduler.go:284
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTask.func1
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  |     /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:347
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  | runtime.goexit
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  |     /usr/local/go/src/runtime/asm_amd64.s:1374
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +Wraps: (3) raft closed timestamp regression in cmd: "\\\xe5\xf8\x16\b\x8e\xfc\xd7" (term: 21, index: 243); batch state: 1623820438.230255107,0?, command: 1623820437.692808095,0, lease: repl=(n1,s1):1 seq=11 start=1623820437.563056021,0 epo=10 pro=1623820437.562975998,0, req: <unknown; not leaseholder>, applying at LAI: 184.
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  | Closed timestamp was set by req: <unknown; not leaseholder or not lease request> under lease: repl=(n9,s9):6 seq=10 start=1623820370.274576462,0 epo=10 pro=1623820370.274576462,0; applied at LAI: 184. Batch idx: 0.
F210616 05:13:58.082972 172 kv/kvserver/store_raft.go:516 ⋮ [n7,s7,r92/3:‹/Table/6{2-3}›] 1013 +  | This assertion will fire again on restart; to ignore run with env var COCKROACH_RAFT_CLOSEDTS_ASSERTIONS_ENABLED=trueRaft log tail:
```

[Full log here.](https://gist.github.com/otan/5014ddf5068a0f04a10e10761c783eba)

Repro: Use https://github.com/cockroachdb/cockroach/compare/master...otan-cockroach:metric_measuring, rebuild roachtest and run `bin/roachtest run tpcc/multiregion/survive=region/chaos=true --debug`. I can reliably hit this within 2 or 3 runs of the roachtest after about 20mins.

**Comment by andreimatei** [comment]
I don't think this issue has much to do with multi-region; removing the respective labels.

@otan I'll take over cluster `otan-1623817213-01-n10cpu4-geo`, which seems to have hit this.

**Comment by andreimatei** [comment]
I believe this issue was fixed on master yesterday by accident by #65624. I'm backporting that to 21.1. I think what was going on here was a lease starting at a timestamp below one that was closed under previous lease, through a scenario similar to the one described in #66562. At the time of the closing, the timestamp in question was a future, synthetic one.
#65624 should fix it, as leases no longer declare that they're closing their own timestamp - and so we're no longer regressing from the previously-closed high timestamp.

I'm doing some runs of Oliver's test to confirm the fix.

**Comment by nvanbenschoten** [comment]
Should we close this? It does sound like it was fixed by https://github.com/cockroachdb/cockroach/pull/65624.
